design/keycloak.x/storage-hot-rod.md

```diff
@@ -0,0 +1,292 @@
+    required string entityVersion = 1;
+    required string id = 2;
+    optional string name = 3;
+    optional string clientTemplateId = 4;
+}
+```
+
+This protobuf definition can be automatically generated from an annotated Java class definition. For setting Protobuf message-specific settings there are `@ProtoField` and `@ProtoDoc` (described in the indexing section) annotations.
+
+```java
+public class InfinispanObjectEntity {
+
+    @ProtoField(number = 1, required = true)
+    public int entityVersion = ModelVersion.VERSION_1.getVersion();
+
+    @ProtoField(number = 2, required = true)
+    public String id;
+
+    @ProtoField(number = 3)
+    @ProtoDoc("@Field(index = Index.YES, store = Store.YES, analyze = Analyze.YES, analyzer = @Analyzer(definition = \"keyword\"))")
+    public String name;
+
+    @ProtoField(number = 4)
+    public String clientTemplateId;
+
+}
+```
+
+Each field has its number. These numbers serve for identification during the marshaling/unmarshaling process.
+
+## Changes in object representations
+
+### Addition of new fields
+
+The advantage of Protobuf definitions is that they are flexible. It is possible to add any number of fields on updates. For example, when we want to introduce a new field named `clientScopeId`, we can do it by adding
+
+```java
+@ProtoField(number = 5)
+public String clientScopeId;
+```
+
+to the message definition and updating the definition in the Infinispan server and in the current `SerializationContext`. Then everything should work as before. It is also possible to read objects with new fields from the old storage version. The new field `clientScopeId` is skipped during unmarshaling.
+
+### Changing field name/type
+
+It is possible to change field names or types. However, such changes can introduce backward incompatibility that we need to avoid due to no-downtime requirements. We can avoid breaking backward compatibility by doing updates in the following steps (in this example, we have the field `clientTemplateId` in the first object version (`entityVersion = 1`) and we want to replace it with `clientScopeId`; the migration path is following `clientScopeId = "template-" + clientTemplateId`):
+
+---
+
+**NOTE**
+
+In this proposal, we distinguish between two types of version:
+
+* **Version of storage** - This corresponds to the version of source code and message definitions on the Keycloak server side; in other words an incremented storage version means there was a release of Keycloak. In a world where there is no change in Keycloak codebase other than changes in `InfinispanObjectEntity`, `entityVersion` corresponds to Keycloak version.
+* **Version of object** - This refers to the version of object in the storage (Infinispan server); in other words, this says what version of storage originally created the object in the store. The object version can be incremented by reading the old object from the store by newer storage, migrating it, and storing it back into the storage.
```

Comment:
I can see your point, and the naming seems to have introduced more issues than anticipated. 
It is a logical version of an object (as it is not tied to any other versioning scheme), but it can be also read as version of a logical object, which is not desired. Let's update the main spec then.